### ISSUE TITLE
type of iterators are not overwritten when name is re-used

### ISSUE THREAD

**Issue by simonmarcin** [python/mypy#13186]
Types of iterators get reused, even when a new iterator of another type gets created with the same name.

**To Reproduce**
```python
import typing

c: typing.List[str] = ['foo']
d: typing.List[int] = [0]

for i in c:
    print(i)

for i in d:
    print(i+1)
```

```
scratch.py:9: error: Incompatible types in assignment (expression has type "int", variable has type "str")
scratch.py:10: error: Unsupported operand types for + ("str" and "int")

```
Same is true for iterators of dicts or dict.items() and so on.

**Expected Behavior**

If iterator gets renamed, everything works as expected.

```python
import typing

c: typing.List[str] = ['foo']
d: typing.List[int] = [0]

for i in c:
    print(i)

for i_new in d:
    print(i_new+1)

```
```
Success: no issues found in 1 source file
```
It should also work with the same iterator name.

**Environment**

- Mypy version used: mypy 0.961 (compiled: yes) (from conda-forge)
- Mypy command-line flags: None
- Mypy configuration options from `mypy.ini` (and other config files): default
- Python version used: 3.9.12 (from conda-forge)
- Operating system and version: Linux 5.10 (Manjaro)

**Comment by JelleZijlstra** [comment]
Mypy treats the first usage of a name as establishing its type. You can change this with the `--allow-redefinition` flag.
